### ISSUE TITLE
Floccus sometimes hangs parts of Firefox UI

### ISSUE THREAD

**Issue by mnalis** [floccusaddon/floccus#929]
### Describe the bug

Sometimes, when Floccus shows blue `<->` on it's icon to indicate it is syncing, firefox will behave strangely with parts of UI "hanged".  Eg. when I try to click on icon of some extension (like floccus or any other: foxyproxy, textarea cache, uBlock origin) it will not open that extension window, but will instead just show a white triangle below extension icon - the triangle that normally expands to extension window). I can click on other existing tabs, and even open a new tab, but I cannot open any real URL in that new tabs - it just stands there. Firefox menus work though, as does `about:Config` internal URL.

First time I closed Firefox to recover, but second time it seems to recover by itself after several minutes (maybe dozen?)

### To Reproduce
Not quite sure. It happened twice so far today, but I've been unable to reproduce it at will. 
Seems to happen only while floccus shows it is syncing.
I have 3 syncs enabled to same nextcloud account but different remote directories: `Bookmarks Toolbar`, `Bookmarks menu`, `sync tabs`.

### Expected behavior
UI continuing to work normally while Floccus is syncing.

### Screenshots
Example hang - as I see it does not open any content for github URL, also the white triangle is shown as I try to open floccus extension window but it won't open it:
![sync_tabs-0a](https://user-images.githubusercontent.com/156656/128551283-21f7166a-a6b8-4599-9101-1d45e6092511.png)

### Desktop
 - OS: Debian Buster GNU/Linux
 - Browser: Firefox
 - Browser Version: 78.12.0esr (64-bit)
 - Floccus version: 4.7.0
 - Floccus sync method: Nextcloud bookmarks

### Server
 - OS: Debian Buster GNU/Linux
 - Nextcloud version:  20.0.11
 - Bookmarks app version: 4.4.0

### Debug log
It is often useful to provide a debug log file along with the issue. You can obtain a debug log of the most recent sync runs by clicking 'Debug log' in the bottom right corner of the floccus panel in your browser.

Note: As your debug log usually contains large parts of your bookmark data, instead of attaching it publicly to the issue here, I recommend uploading it privately at the following URL https://cloud.marcelklehr.de/s/PCi3iLWrzdMZSmw

- [ ] Debug log provided (not provided as I cannot click on floccus icon to retrieve it due to nature of the bug. Is there a way to manually retrieve the log outside of firefox?)

### Additional context
I never noticed this problem in previous Floccus `v4.6.4`, but `Sync tabs` **was disabled then** too (because of https://github.com/floccusaddon/floccus/issues/883 and https://github.com/floccusaddon/floccus/issues/912)

Might be related or not: 
- first time I synced tabs after upgrading floccus, after few dozen minutes it started closing and reopening what seems like all tabs (including tabs   I'm sure I haven't changed since initial tabs sync attempts, eg. for  months).  Each one was in "loading" state (usually my firefox is setup to only try loading tabs once I click on them, and not to load them all automatically on eg. restart), and since I have several hundred of tabs it slowed firefox considerably, so I went to do other things on the computer while all those tabs load.

- However some 20 minutes after I stopped watching firefox (so 40-50 minutes after a sync start), I've noticed general sluggishness, and was surprised to see Firefox was still syncing and has eaten all 8GB of RAM *and* 8GB of swap.  I was soon forced to click `x` to close Firefox, which worked.  After that I opened and synced tabs manually and after some time this bug occurred.

- After I restarted firefox and synced tabs again, it worked and said tabs are synced successfully.

- Some time after that, while it was syncing automatically (or it might have been me playing with manual sync before and sync not being completed yet, I'm not sure), floccus experienced the bug again - this time it recovered by itself after some dozen minutes or so.

**Comment by TeutonJon78** [comment]
Not sure if I'm essentially running into the same problem, or if it overlaps with #1011. I have a decent number of bookmarks (probably into the thousands over the years of collecting them). But every time Floccus goes to sync (4.10.1, FF 97.0.1, so all up to date), it basically completely stalls out Firefox for about 40-60 seconds. I can open tabs and scroll existing pages, but it basically kills the internet connection side. No pages will load. After it finishes some part of the sync, the connection will start working again. It will be fine until the next sync starts.

It's not a problem with the internet connection in general, because I can still use other browsers or internet things at the same time and they continue to work just fine.

**Comment by marcelklehr** [comment]
I may have been able to improve this situation with #1053 (by not constructing a hash tree for the whole bookmarks tree anymore).

**Comment by marcelklehr** [comment]
I'd be grateful, if you could help test whether this makes a difference. I'll provide you with an alpha build.

**Comment by pawlosck** [comment]
It looks like similar to my problem. I am using floccus and nextcloud. When floccus starts syncing, firefox is freezed and I can't do anything with firefox. Firefox consumes over 100% of one core.
Today, I left my laptop for few hours, to check, what will be happen and after 3-4h, I saw a lot of messages like this "E017: Network error: Check your network connection and your account details".
Bookmarks was saved in nextcloud, so generally is ok, but freezed firefox and one core consumed by firefox is not good.
I checked apache logs, nextcloudd logs and floccus logs and I didn't see any strange logs except problems with internet connection reported by floccus.

**Comment by marcelklehr** [comment]
> I'd be happy to try the test build.


Here you go:
I've already had one report of it not working, sadly, but I'm always glad about more testers.
https://github.com/floccusaddon/floccus/releases/tag/v4.11.0-alpha.0

**Comment by pawlosck** [comment]
Is it stable? I mean, is it safe for my bookmarks? Alpha versions is not stable in most of projects.

**Comment by marcelklehr** [comment]
It [passed all automated tests](https://github.com/floccusaddon/floccus/runs/5376713377) at the time, but I generally recommend making backups when using floccus.

**Comment by TeutonJon78** [comment]
Is there any way to install it on normal Firefox? It seems unsigned addons can only be installed in daily or dev builds.

**Comment by marcelklehr** [comment]
You may be able to install it at about:debugging -> This firefox -> Load temporary addon

**Comment by exstntlmsnthrp** [comment]
I agree with @TeutonJon78, this seems at minimum very closely related to [issue 1011](https://github.com/floccusaddon/floccus/issues/1011), which I replied to before finding this. As I commented there, I have most of the same issues referenced in this issue, except with Nextcloud.

I have done similar to what @pawlosck says in their comment as well. Although, for me Firefox will crash before it gets that long.
